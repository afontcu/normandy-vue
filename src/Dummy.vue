<template>
  <div class="Dummy"></div>
</template>

<script>
export default {
  name: 'Dummy'
}
</script>
