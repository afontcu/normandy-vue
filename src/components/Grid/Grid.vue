<template>
  <div class="o-grid" :class="[ classes ]">
    <slot></slot>
  </div>
</template>

<script>
import VueTypes from 'vue-types'
import { VALIGN_VALUES, SIZE_VALUES, SPACE_VALUES } from '@/constraints'

export default {
  name: 'Grid',
  props: {
    valign: VueTypes.oneOf(VALIGN_VALUES),
    size: VueTypes.oneOf(SIZE_VALUES),
    space: VueTypes.oneOf(SPACE_VALUES),
    stretch: VueTypes.bool.def(false),
    auto: VueTypes.bool.def(false),
    liquid: VueTypes.bool.def(false),
    reverse: VueTypes.bool.def(false),
    wrapped: VueTypes.bool.def(false)
  },
  data () {
    return {}
  },
  computed: {
    classes () {
      return {
        [`o-grid--${this.valign}`]: this.valign,
        [`o-grid--${this.size}`]: this.size,
        [`o-grid--${this.space}`]: this.space,
        'o-grid--stretch': this.stretch,
        'o-grid--wrapped': this.wrapped,
        'o-grid--liquid': this.liquid,
        'o-grid--auto o-grid--liquid': this.auto,
        'o-grid--reverse': this.reverse

      }
    }
  }
}
</script>
