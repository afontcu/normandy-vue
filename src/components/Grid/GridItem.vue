<template>
  <div class="o-grid__item" :style="computedStyle">
    <slot></slot>
  </div>
</template>

<script>
export default {
  name: 'Grid-item',
  data () {
    return {}
  },
  props: {
    width: {
      type: String,
      validator: value => {
        const split = value.split('/')
        return value.match(/^[0-9]\/[0-9]$/g) && split[0] <= split[1]
      }
    }
  },
  computed: {
    computedStyle () {
      if (!this.width) return null

      return { width: this.getWidthFromFraction(this.width) }
    }
  },
  methods: {
    getWidthFromFraction (fraction) {
      const split = fraction.split('/')
      const result = parseInt(split[0]) / parseInt(split[1])

      return this.round(result, 5) * 100 + '%'
    },
    round (value, decimals) {
      return Number(Math.round(value + 'e' + decimals) + 'e-' + decimals).toFixed(decimals)
    }
  }
}
</script>
