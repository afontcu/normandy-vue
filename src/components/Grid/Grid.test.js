import Grid from './Grid'
import { mount } from 'avoriaz'
import Dummy from '@/Dummy'

describe('Grid.vue', () => {
  beforeAll(() => console.error = jest.fn(error => {
    throw new Error(error)
  }))

  it('renders correct nodes when required props are provided', () => {
    const wrapper = mount(Grid)
    expect(wrapper.is('div')).toBe(true)
    expect(wrapper.hasClass('o-grid')).toBe(true)
  })

  it('renders slot content when defined', () => {
    const wrapper = mount(Grid, { slots: { default: Dummy } })
    expect(wrapper.first('.Dummy')).toBeTruthy()
  })

  it('renders an snapshot', () => {
    const wrapper = mount(Grid)
    expect(wrapper.html()).toMatchSnapshot('default')
  })

  it('adds align class when align prop is provided', () => {
    const values = ['bottom', 'middle']

    values.forEach((align) => {
      const wrapper = mount(Grid, { propsData: { align } })
      expect(wrapper.hasClass(`o-grid--${align}`)).toBe(true)
    })
  })

  it('adds size class when size prop is provided', () => {
    const values = ['flush', 'tiny', 'small', 'large', 'huge']

    values.forEach((size) => {
      const wrapper = mount(Grid, { propsData: { size } })
      expect(wrapper.hasClass(`o-grid--${size}`)).toBe(true)
    })
  })

  it('adds space class when space prop is provided', () => {
    const values = ['around', 'between', 'center', 'end']

    values.forEach((space) => {
      const wrapper = mount(Grid, { propsData: { space } })
      expect(wrapper.hasClass(`o-grid--${space}`)).toBe(true)
    })
  })

  it('adds reverse class when reverse prop is provided', () => {
    const wrapper = mount(Grid, { propsData: { reverse: true } })
    expect(wrapper.hasClass('o-grid--reverse')).toBe(true)
  })

  it('adds stretch class when stretch prop is provided', () => {
    const wrapper = mount(Grid, { propsData: { stretch: true } })
    expect(wrapper.hasClass('o-grid--stretch')).toBe(true)
  })

  it('adds wrapped class when wrapped prop is provided', () => {
    const wrapper = mount(Grid, { propsData: { wrapped: true } })
    expect(wrapper.hasClass('o-grid--wrapped')).toBe(true)
  })

  it('adds liquid class when liquid prop is provided', () => {
    const wrapper = mount(Grid, { propsData: { liquid: true } })
    expect(wrapper.hasClass('o-grid--liquid')).toBe(true)
  })

  it('adds liquid and auto class when auto prop is provided', () => {
    const wrapper = mount(Grid, { propsData: { auto: true } })
    expect(wrapper.hasClass('o-grid--auto')).toBe(true)
  })

  it('adds all optional classes when all props are provided', () => {
    const wrapper = mount(Grid, {
      propsData: {
        size: 'small',
        align: 'middle',
        space: 'end',
        reverse: true,
        stretch: true,
        liquid: true,
        auto: true,
        wrapped: true
      }
    })
    expect(wrapper.hasClass('o-grid--small')).toBe(true)
    expect(wrapper.hasClass('o-grid--middle')).toBe(true)
    expect(wrapper.hasClass('o-grid--end')).toBe(true)
    expect(wrapper.hasClass('o-grid--reverse')).toBe(true)
    expect(wrapper.hasClass('o-grid--stretch')).toBe(true)
    expect(wrapper.hasClass('o-grid--liquid')).toBe(true)
    expect(wrapper.hasClass('o-grid--auto')).toBe(true)
    expect(wrapper.hasClass('o-grid--wrapped')).toBe(true)
  })

  it('changes size prop when its class', () => {
    const wrapper = mount(Grid, { propsData: { size: 'large' } })
    expect(wrapper.hasClass('o-grid--large')).toBe(true)
    expect(wrapper.hasClass('o-grid--small')).toBe(false)

    wrapper.setProps({ size: 'small' })

    expect(wrapper.hasClass('o-grid--small')).toBe(true)
    expect(wrapper.hasClass('o-grid--large')).toBe(false)
  })

  it('changes align prop when its class', () => {
    const wrapper = mount(Grid, { propsData: { align: 'bottom' } })
    expect(wrapper.hasClass('o-grid--bottom')).toBe(true)
    expect(wrapper.hasClass('o-grid--middle')).toBe(false)

    wrapper.setProps({ align: 'middle' })

    expect(wrapper.hasClass('o-grid--middle')).toBe(true)
    expect(wrapper.hasClass('o-grid--bottom')).toBe(false)
  })

  it('changes space prop when its class', () => {
    const wrapper = mount(Grid, { propsData: { space: 'center' } })
    expect(wrapper.hasClass('o-grid--center')).toBe(true)
    expect(wrapper.hasClass('o-grid--around')).toBe(false)

    wrapper.setProps({ space: 'around' })

    expect(wrapper.hasClass('o-grid--around')).toBe(true)
    expect(wrapper.hasClass('o-grid--center')).toBe(false)
  })

  it('breaks with invalid size prop value', () => {
    expect(() => {
      mount(Grid, { propsData: { size: 'willfail' } })
    }).toThrow()
  })

  it('breaks with invalid align prop value', () => {
    expect(() => {
      mount(Grid, { propsData: { align: 'willfail' } })
    }).toThrow()
  })

  it('breaks with invalid space prop value', () => {
    expect(() => {
      mount(Grid, { propsData: { space: 'willfail' } })
    }).toThrow()
  })

  it('renders an snapshot with modifier class when size prop is provided', () => {
    const wrapper = mount(Grid, {
      slots: { default: Dummy },
      propsData: { size: 'large' }
    })
    expect(wrapper.html()).toMatchSnapshot('size')
  })

  it('renders an snapshot with modifier class when align prop is provided', () => {
    const wrapper = mount(Grid, {
      slots: { default: Dummy },
      propsData: { align: 'middle' }
    })
    expect(wrapper.html()).toMatchSnapshot('align')
  })

  it('renders an snapshot with modifier class when space prop is provided', () => {
    const wrapper = mount(Grid, {
      slots: { default: Dummy },
      propsData: { space: 'between' }
    })
    expect(wrapper.html()).toMatchSnapshot('space')
  })

  it('renders an snapshot with modifier class when stretch prop is provided', () => {
    const wrapper = mount(Grid, {
      slots: { default: Dummy },
      propsData: { stretch: true }
    })
    expect(wrapper.html()).toMatchSnapshot('stretch')
  })

  it('renders an snapshot with modifier class when wrapped prop is provided', () => {
    const wrapper = mount(Grid, {
      slots: { default: Dummy },
      propsData: { wrapped: true }
    })
    expect(wrapper.html()).toMatchSnapshot('wrapped')
  })

  it('renders an snapshot with modifier class when liquid prop is provided', () => {
    const wrapper = mount(Grid, {
      slots: { default: Dummy },
      propsData: { liquid: true }
    })
    expect(wrapper.html()).toMatchSnapshot('liquid')
  })

  it('renders an snapshot with modifier class when auto prop is provided', () => {
    const wrapper = mount(Grid, {
      slots: { default: Dummy },
      propsData: { auto: true }
    })
    expect(wrapper.html()).toMatchSnapshot('auto')
  })

  it('renders an snapshot with modifier class when reverse prop is provided', () => {
    const wrapper = mount(Grid, {
      slots: { default: Dummy },
      propsData: { reverse: true }
    })
    expect(wrapper.html()).toMatchSnapshot('reverse')
  })
})
