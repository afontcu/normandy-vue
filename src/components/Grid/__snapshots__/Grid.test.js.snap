// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auto 1`] = `

<div class="o-grid o-grid--auto o-grid--liquid">
  <div class="Dummy">
  </div>
</div>

`;

exports[`default 1`] = `

<div class="o-grid">
</div>

`;

exports[`liquid 1`] = `

<div class="o-grid o-grid--liquid">
  <div class="Dummy">
  </div>
</div>

`;

exports[`reverse 1`] = `

<div class="o-grid o-grid--reverse">
  <div class="Dummy">
  </div>
</div>

`;

exports[`size 1`] = `

<div class="o-grid o-grid--large">
  <div class="Dummy">
  </div>
</div>

`;

exports[`space 1`] = `

<div class="o-grid o-grid--between">
  <div class="Dummy">
  </div>
</div>

`;

exports[`stretch 1`] = `

<div class="o-grid o-grid--stretch">
  <div class="Dummy">
  </div>
</div>

`;

exports[`valign 1`] = `

<div class="o-grid o-grid--middle">
  <div class="Dummy">
  </div>
</div>

`;

exports[`wrapped 1`] = `

<div class="o-grid o-grid--wrapped">
  <div class="Dummy">
  </div>
</div>

`;
